File: python/easy/min_stack.py
class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.elem_stack = list()
        self.min_stack = list()
        
    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        self.elem_stack.append(x)
        if len(self.min_stack) == 0:
            self.min_stack.append(x)
        elif self.min_stack[len(self.min_stack)-1] >= x:
            self.min_stack.append(x)

    def pop(self):
        """
        :rtype: void
        """
        if len(self.elem_stack) == 0:
            return
        top_val = self.elem_stack[len(self.elem_stack)-1]
        if top_val == self.min_stack[len(self.min_stack)-1]:
            self.min_stack.pop()
        self.elem_stack.pop()
        

    def top(self):
        """
        :rtype: int
        """
        return self.elem_stack[len(self.elem_stack)-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.min_stack[len(self.min_stack)-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: cpp/easy/min_stack.cpp
class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() {
        
    }
    
    void push(int x) {
        if(ret.empty())
        {
            ret.push_back(x);
            minSt.push_back(x);
        }
        else
        {
            ret.push_back(x);
            if(x <= minSt.back())
                minSt.push_back(x);
        }
    }
    
    void pop() {
        if(ret.empty())
            return;
        int top_val = ret.back();
        if(top_val == minSt.back())
            minSt.pop_back();
        ret.pop_back();
    }
    
    int top() {
        if(ret.empty())
            return -1;
        int top_val = ret.back();
        return top_val;
    }
    
    int getMin() {
        
        return minSt.back();
    }

private:
    vector<int> ret;
    vector<int> minSt;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */


Conclusion: an easy algorithm issue; actually, it tests the data structure of stack, we can use two vectors to implement in c++

Time Complexity: O(n)
Space Complexity:O(n)
